api/tokens.go

```diff
@@ -148,3 +151,72 @@ func (api *API) RefreshHandler(ctx context.Context, w http.ResponseWriter, req *
+func (api *API) SignOutAllUsersHandler(ctx context.Context, w http.ResponseWriter, req *http.Request) (*models.SuccessResponse, *models.ErrorResponse) {
+	usersList := models.UsersList{}
+	listUserInput := usersList.BuildListUserRequest("status=\"Enabled\"", "username", int64(0), &api.UserPoolId)
+	listUserResp, err := api.CognitoClient.ListUsers(listUserInput)
+	if err != nil {
+		return nil, models.NewErrorResponse(http.StatusInternalServerError, nil, err)
+	}
+	usersList.MapCognitoUsers(listUserResp)
+	globalSignOut := &models.GlobalSignOut{
+		ResultsChannel: make(chan string, len(usersList.Users)),
+		BackoffSchedule: []time.Duration{
+			1 * time.Second,
+			3 * time.Second,
+			10 * time.Second,
+		},
+		RetryAllowed: true,
+	}
+	// run api.SignOutUsersWorker concurrently
+	go api.SignOutUsersWorker(ctx, globalSignOut, &usersList.Users)
+	return models.NewSuccessResponse(nil, http.StatusAccepted, nil), nil
+}
+
+// SignOutUsersWorker - signs out users globally by invalidating user's refresh token
+func (api *API) SignOutUsersWorker(ctx context.Context, g *models.GlobalSignOut, usersList *[]models.UserParams) {
+	userSignOutRequestData := g.BuildSignOutUserRequest(usersList, &api.UserPoolId)
+	for i := range userSignOutRequestData {
+		for _, backoff := range g.BackoffSchedule {
+			_, err := api.generateGlobalSignOutRequest(userSignOutRequestData[i])
+			if err != nil {
+				responseErr := models.NewCognitoError(ctx, err, "Cognito AdminUserGlobalSignOut request for sign out")
+				if responseErr.Code != models.TooManyRequestsError { // 1. Process all errors other than TooManyRequestsException (429)
+					if g.RetryAllowed { // 2. Attempt one more request to AdminUserGlobalSignOut if GlobalSignOut.RetryAllowed true, else break to next user
+						g.RetryAllowed = false // 3. Set GlobalSignOut.RetryAllowed to false and request AdminUserGlobalSignOut
+						log.Event(ctx, "Error Cognito AdminUserGlobalSignOut:", log.ERROR, log.Error(err))
+						_, retryErr := api.generateGlobalSignOutRequest(userSignOutRequestData[i])
+						if retryErr != nil { // 4. If error response from request received, process it
+							retryResponseErr := models.NewCognitoError(ctx, err, "Cognito AdminUserGlobalSignOut request for sign out")
+							if retryResponseErr.Code != models.TooManyRequestsError { // 4.1 If error is not a TooManyRequestsException (429), reset GlobalSignOut.RetryAllowed and break to next user
+								g.RetryAllowed = true
+								break
+							}
+						} else { // 5. If request successful, reset GlobalSignOut.RetryAllowed and set original response error to nil - username will be added to GlobalSignOut.ResultsChannel
+							g.RetryAllowed = true
+							err = nil
+						}
+					} else { // 6. If GlobalSignOut.RetryAllowed is already false, know that a second request already made, so reset GlobalSignOut.RetryAllowed and break to next user 
+						g.RetryAllowed = true
+						break
+					}
+				}
+			}
+			if err == nil {
```

Comment:
Instead of an `if err != nil` clause followed by a `if err == nil` clause, it would probably be better to use an `else` clause

Reply:
I think this is an `if` rather than an `else` so if the retry succeeds `err` is reset to `nil` on line 199 and this `if` can then catch both.
But i think it might be sacrificing readability to not repeat 2 lines.

Reordering the if check and repeating the lines would remove a layer of nesting and might improve readability

```suggestion
if err == nil {
	g.ResultsChannel <- *userSignOutRequestData[i].Username
	break
}
responseErr := models.NewCognitoError(ctx, err, "Cognito AdminUserGlobalSignOut request for sign out")
if responseErr.Code != models.TooManyRequestsError {
	if g.RetryAllowed { 
		g.RetryAllowed = false
		_, retryErr := api.generateGlobalSignOutRequest(userSignOutRequestData[i])
		if retryErr != nil { 
			retryResponseErr := models.NewCognitoError(ctx, err, "Cognito AdminUserGlobalSignOut request for sign out")
			if retryResponseErr.Code != models.TooManyRequestsError { 
				g.RetryAllowed = true
				break
			}
		} else { 
			g.RetryAllowed = true
			g.ResultsChannel <- *userSignOutRequestData[i].Username
			break
		}
	} else { 
		g.RetryAllowed = true
		break
	}
}
time.Sleep(backoff)
```